###################################################################
###################################################################
####
#### after running GEMMA using the commands generated by script 3.01_input_GWAS_SNPs (GEMMA_commands_SNPs.sh)
#### run this script for each trait
####
#### for each environment:
#### - generate manhattan plot
#### - generate QQ plot
#### - write list of significant SNPs
####
#### Manfred Mayer (Technical University of Munich, Plant Breeding)
#### manfred.mayer@tum.de
####
#### date: 29.06.2020
###################################################################
###################################################################

# general settings
options(stringsAsFactors=FALSE)
options(scipen=99)
options(warn = 1)
set.seed(212)

# arguments
TRAIT <- "PH_final"

args=(commandArgs(TRUE))
if(length(args)==0){
  print("No arguments supplied.")
}else{
  for(i in 1:length(args)){
    eval(parse(text=args[[i]]))
  }
}
TRAIT

# functions
# plot function for manhattan plots
manhat_plot_f <- function(res, lwd = 3){
	points(x = c(res[1], res[2]), y = c(res[3], res[3]), type = "l", lwd = lwd, col = res[4])
	}

# colors
cols_manhattan <- c(rgb(125/255, 125/255, 125/255, alpha = 0.9), rgb(5/255, 5/255, 5/255, alpha = 0.9))
cols_manhattan

# define centromere positions and chromosome sizes (B73v4)
Start_bp <- c(136.77, 95.51, 85.78, 109.07, 104.54, 52.3, 56.38, 50.53, 53.75, 51.39)
End_bp <- c(137.12, 97.49, 86.93, 110.5, 106.82, 53.11, 56.68, 52.07, 55.39, 52.78)
centromers <- cbind(Start_bp, End_bp)
chr_size <- c(307.041717, 244.442276, 235.667834, 246.994605, 223.902240, 174.033170, 182.381542, 181.122637, 159.769782, 150.982314)
chr_size_temp <- c(0, chr_size)*1000000
chr_axis_at <- NULL
for(CHR in 1:10){
chr_axis_at <- c(chr_axis_at, mean(c(sum(chr_size_temp[1:CHR]), sum(chr_size_temp[1:(CHR+1)]))))
}

###################################################################
###
### 
###

# output folder
dir.create(paste("GWAS_SNPs/", TRAIT, "/summary_GWAS", sep =""))
outfolder <- paste("GWAS_SNPs/", TRAIT, "/summary_GWAS", sep ="")

# graphical parameters
cex <- 0.6
cex.axis <- 1.6
cex.lab <- 1.6
	
# load data
# load non-collinear markers
load(paste("GWAS_SNPs/", TRAIT, "/DHs_SNPs_.nonColl_list.RData", sep = ""))
str(nonColl_list)

# geno and map
load(paste("geno_InfoList_DH_SNPs.RData", sep = ""))
str(InfoList)
str(geno)

Envs <- read.table(paste0("GWAS_SNPs/", TRAIT, "/", list.files(path = paste0("GWAS_SNPs/", TRAIT), pattern = ".phenotypes_order.txt")), stringsAsFactors = FALSE)
Envs <- Envs[,1]
Envs

# run for each environment (and for the across-environment-BLUEs)
for (Env_i in 1:length(Envs)){
ENV <- Envs[Env_i]
print(ENV)

path <- paste("GWAS_SNPs/", TRAIT, "/output", sep ="")

	## load GWAS results
	if(file.exists(paste(path, "/", TRAIT, "_", ENV, ".assoc.txt", sep =""))){
		system(paste("gzip ", file.path(path, paste(TRAIT, "_", ENV, ".assoc.txt", sep = "")), sep =""))
	}
	GWASresults <- read.table(paste(path, "/", TRAIT, "_", ENV, ".assoc.txt.gz", sep =""), header = TRUE)	

	## load again phenotypic data and calculate environment-specific standard deviation
	pheno_trait <- read.table(paste("GWAS_SNPs/", TRAIT, "/", list.files(path = paste0("GWAS_SNPs/", TRAIT), pattern = ".pheno.txt"), sep =""), header = FALSE)
	pheno_trait <- pheno_trait[,Env_i]
	sd_pheno <- sd(pheno_trait, na.rm = TRUE)
	print(paste("sd_pheno", sd_pheno))

	## calculate FDR-threshold
	# take non-collinear markers into account
	nonColl_marker <- nonColl_list[[ENV]]
	GWASresults_nonColl <- GWASresults[which(GWASresults$rs %in% nonColl_marker), ]
	# adjust p-values according to FDR (Benjamini and Hochberg)
	p.adjusted_BH <- p.adjust(GWASresults_nonColl$p_lrt, method = "BH")
	# calculate significance threshold
	FDR0.15 <- ((length(which(p.adjusted_BH < 0.15)) + 1) / length(GWASresults_nonColl$p_lrt)) * 0.15
	print(paste("-log10 FDR0.15", -log10(FDR0.15)))
	
	# define trait-associated haplotypes
	# p-values were calculated for all markers, also including haplotypes with r2=1, as these need to be considered for defining the QTL regions in a later step
	hits_FDR0.15 <- GWASresults$rs[which(GWASresults$p_lrt <= FDR0.15)]
	print(paste("FDR0.15:", length(hits_FDR0.15), "sig. markers"))

	## generate Manhattan-plot
	if(length(hits_FDR0.15) == 0){
	highlight <- NULL
	} else {
	highlight <- hits_FDR0.15
	}
	# highlight all markers with a significant favorable/unfavorable effect
	if(gsub("BLUE.", "", TRAIT) %in% gsub("BLUE.", "", c(	'EV_V4', 'EV_V6',
															'PH_V4', 'PH_V6',
															'PH_final'))){	
		highlight_fav <- GWASresults$rs[which(GWASresults$beta > 0)]	
	} else {
	#MF, FF, LO, TILL
		highlight_fav <- GWASresults$rs[which(GWASresults$beta < 0)]	
	}	
	highlight_fav <- highlight_fav[which(highlight_fav %in% hits_FDR0.15)]
	highlight_unfav <- highlight[which((highlight %in% highlight_fav) == FALSE)]
	
	info_list_temp <- InfoList[GWASresults$rs, ]
	GWASresults$Start_bp <- info_list_temp[,2]
	GWASresults$End_bp <- info_list_temp[,3]

	#
	# this section is for plotting every haplotype as a line from it's start to it's end position
	# however, the plotting can take a while (especially on local desktop machines), therefore it is not run here
	# it only matters when you zoom into specific regions, in a genome-wide view all haplotypes appear more or less like a dot anyways
	#
	#png(paste(outfolder, "/", TRAIT, "_", ENV, "_Manhattan_hap.png", sep =""), width = 3000, height = 1400, res = 200)
	#par(mar = c(5, 5, 4, 2) + 0.1)
	#res <- NULL
	#cols <- cols_manhattan
	#col_i <- 2
	#for(CHR in sort(unique(GWASresults$chr))){
	#col_i <- ifelse(col_i == 2, 1, 2)
	#temp <- GWASresults[which(GWASresults$chr == CHR), ]
	#rownames(temp) <- temp$rs
	#temp <- temp[ , c("Start_bp", "End_bp", "p_lrt")]
	#temp$Start_bp <- temp$Start_bp + sum(chr_size_temp[1:CHR])
	#temp$End_bp <- temp$End_bp + sum(chr_size_temp[1:CHR])
	#temp$p_lrt <- -log10(temp$p_lrt)
	#temp$cols <- rep(cols[col_i], nrow(temp))
	#res <- rbind(res, temp)
	#rm(temp)
	#}
	#plot(x = -100, y = -100, xlim = c(0, sum(chr_size_temp)), ylim = c(0, max(c(res$p_lrt, -log10(c(FDR0.15)))*1.04)), main = paste(TRAIT, "_", ENV, " Manhattan (FDR 15%)"), ylab = expression(-log[10](italic(p))), xlab = "Chromosome", xaxt = "n", cex.axis = cex.axis, cex.lab = cex.lab, cex = cex)
	#
	#abline(h = -log10(c(FDR0.15)), col = c("darkorange2"), lwd = 1.3)
	#legend("topright", lwd = 2, lty = 1, col = c("darkorange2"), legend = c("FDR 15%"), cex = 0.9, bty = "n")
	#axis(side = 1, at = chr_axis_at, labels = 1:10, cex.axis = cex.axis, cex.lab = cex.lab, cex = cex)
	#apply(res, 1, manhat_plot_f, lwd = 3)	
	#if (length(highlight_unfav) > 0) {
	#	res[highlight_unfav, "cols"] <- rgb(0.9, 0, 0, alpha = 0.75)
	#	apply(res[highlight_unfav, ], 1, manhat_plot_f, lwd = 4.5)
	#}
	#if (length(highlight_fav) > 0) {
	#	res[highlight_fav, "cols"] <- rgb(0.1, 0.45, 1.00, alpha = 0.95)
	#	apply(res[highlight_fav, ], 1, manhat_plot_f, lwd = 4.5)
	#}	
	#dev.off()

	png(paste(outfolder, "/", TRAIT, "_", ENV, "_Manhattan.png", sep =""), width = 3000, height = 1400, res = 200)
	par(mar = c(5, 5, 4, 2) + 0.1)
	res <- NULL
	cols <- cols_manhattan
	col_i <- 2
	for(CHR in sort(unique(GWASresults$chr))){
	  col_i <- ifelse(col_i == 2, 1, 2)
	  temp <- GWASresults[which(GWASresults$chr == CHR), ]
	  rownames(temp) <- temp$rs
	  temp <- temp[ , c("Start_bp", "End_bp", "p_lrt")]
	  temp$Start_bp <- temp$Start_bp + sum(chr_size_temp[1:CHR])
	  temp$End_bp <- temp$End_bp + sum(chr_size_temp[1:CHR])
	  temp$p_lrt <- -log10(temp$p_lrt)
	  temp$cols <- rep(cols[col_i], nrow(temp))
	  res <- rbind(res, temp)
	  rm(temp)
	}
	plot(x = -100, y = -100, xlim = c(0, sum(chr_size_temp)), ylim = c(0, max(c(res$p_lrt, -log10(c(FDR0.15)))*1.04)), main = paste(TRAIT, "_", ENV, " Manhattan (FDR 15%)"), ylab = expression(-log[10](italic(p))), xlab = "Chromosome", xaxt = "n", cex.axis = cex.axis, cex.lab = cex.lab, cex = cex)

	abline(h = -log10(c(FDR0.15)), col = c("darkorange2"), lwd = 1.3)
	legend("topright", lwd = 2, lty = 1, col = c("darkorange2"), legend = c("FDR 15%"), cex = 0.9, bty = "n")
	axis(side = 1, at = chr_axis_at, labels = 1:10, cex.axis = cex.axis, cex.lab = cex.lab, cex = cex)
	points(x = apply(res[ , c(1,2)], 1, mean), y = res$p_lrt, pch = 20, col = res$cols, cex = cex)
	if (length(highlight_unfav) > 0) {
	  res[highlight_unfav, "cols"] <- rgb(0.9, 0, 0, alpha = 0.75)
	  apply(res[highlight_unfav, ], 1, manhat_plot_f, lwd = 4.5)
	}
	if (length(highlight_fav) > 0) {
	  res[highlight_fav, "cols"] <- rgb(0.1, 0.45, 1.00, alpha = 0.95)
	  apply(res[highlight_fav, ], 1, manhat_plot_f, lwd = 4.5)
	}	
	dev.off()

	# generate QQ-plot
	png(paste(outfolder, "/", TRAIT, "_", ENV, "_QQ.png", sep =""), width = 1400, height = 1400, res = 200)
	par(mar = c(5, 4, 4, 2) + 0.1)
	res <- GWASresults_nonColl
	pvector <- res$p_lrt
	names(pvector) <- res$rs
	o = -log10(sort(pvector, decreasing = FALSE))
	e = -log10(ppoints(length(pvector)))
	plot(x = e, y = o, pch = 1, xlim = c(0, max(e)), ylim = c(0, max(o)), xlab = expression(Expected ~ ~-log[10](italic(p))),
		 ylab = expression(Observed ~ ~-log[10](italic(p))), col = cols_manhattan[2], lwd = 1.5, main = paste(TRAIT, ENV, "QQ"))
	abline(0, 1, col = "red")
	dev.off()
	
	## write out significant haplotypes
	rownames(GWASresults) <- GWASresults$rs
	Effects <- GWASresults$beta
	GWASresults <- cbind(GWASresults, Effects)
	
	# write out each hit
		if(length(highlight) != 0){
		regions <- GWASresults[highlight, c("chr", "Start_bp", "End_bp", "af", "p_lrt", "Effects")]
		regions <- regions[order(regions[,1], regions[,2], regions[,3]),]
		regions <- cbind(regions, regions[,6] / sd_pheno)
		colnames(regions) <- c("Chr", "Start_bp", "End_bp", "allele_freq", "pvalue", "effect", "effect_prop_sd")
		write.table(regions, file = paste(outfolder, "/Hits_", TRAIT, "_", ENV, "_FDR0.15.csv", sep =""), sep = ";" , dec=".", row.names = T, col.names = NA)
		} else {
		regions <- matrix(rep(NA, 7), ncol = 7)
		colnames(regions) <- c("Chr", "Start_bp", "End_bp", "maf", "pvalue", "effect", "effect_prop_sd")
		write.table(regions, file = paste(outfolder, "/Hits_", TRAIT, "_", ENV, "_FDR0.15.csv", sep =""), sep = ";" , dec=".", row.names = T, col.names = NA)
		}

}

###
###################################################################
